### PR TITLE
Refactor the internal http client

### PR DESCRIPTION
**Pull request by thunderboltsid**
Refactor the client with:
- Modify NewClient constructor to use functional options
- remove NewBaseClient constructor
- Add logr based configurable logging

### CODE REVIEW

**Comment by thunderboltsid** [comment]
Turns out Foundation APIs need http. Will add explicit option for adding scheme selection.

**Comment by thunderboltsid** [comment]
> Turns out Foundation APIs need http. Will add explicit option for adding scheme selection.

On another thought, that can be done by just setting the WithBaseURL with a `http://`. so, no `WithScheme` option.

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,11 @@ go 1.16
 
 require (
 	github.com/PaesslerAG/jsonpath v0.1.1
-	github.com/hashicorp/terraform-plugin-sdk/v2 v2.17.0
+	github.com/go-logr/logr v1.2.3
```
Comment (heiko-koehler) [inline go.mod:7]: Neat! No more dependency on terraform. We use zap in CNDS as well as part of controller-runtime

File: internal/client.go [hunk 2]
```diff
@@ -16,152 +16,207 @@ import (
...
+		}
+
+		c.cookies = resp.Cookies()
+	}
 
 	return c, nil
 }
 
 // NewRequest creates a request
 func (c *Client) NewRequest(ctx context.Context, method, urlStr string, body interface{}) (*http.Request, error) {
```
Comment (heiko-koehler) [inline internal/client.go:211]: Is context used here at all?
We should allocate http request using this context to propagate timeout.
See https://pkg.go.dev/net/http#NewRequestWithContext

File: internal/client.go [hunk 3]
```diff
@@ -16,152 +16,207 @@ import (
...
+	c.httpClient.Transport = transCfg
 
-	baseURL, err := url.Parse(fmt.Sprintf(defaultBaseURL, protocol, credentials.URL))
-	if err != nil {
-		return nil, err
-	}
+	if c.credentials.SessionAuth {
+		c.logger.V(1).Info("Using session_auth")
+
+		ctx := context.TODO()
```
Comment (heiko-koehler) [inline internal/client.go:188]: Can we at least create new context with some very high timeout like 1 minute? I'm thinking of cases where server stopped responding but is still reachable via network.

File: internal/client.go [hunk 4]
```diff
@@ -182,26 +237,26 @@ func (c *Client) NewRequest(ctx context.Context, method, urlStr string, body int
...
 		req.Header.Add("Authorization", "Basic "+
-			base64.StdEncoding.EncodeToString([]byte(c.Credentials.Username+":"+c.Credentials.Password)))
+			base64.StdEncoding.EncodeToString([]byte(c.credentials.Username+":"+c.credentials.Password)))
 	}
 	return req, nil
 }
 
-// NewRequest creates a request without authorisation headers
+// NewUnAuthRequest creates a request without authorisation headers
 func (c *Client) NewUnAuthRequest(ctx context.Context, method, urlStr string, body interface{}) (*http.Request, error) {
```
Comment (heiko-koehler) [inline internal/client.go:252]: Similar to above. context is not used by request.

File: internal/client.go [hunk 5]
```diff
@@ -342,13 +397,13 @@ func (c *Client) OnRequestCompleted(rc RequestCompletionCallback) {
...
 // Do performs request passed
 func (c *Client) Do(ctx context.Context, req *http.Request, v interface{}) error {
-	// check if client exists or not
-	if c.client == nil {
+	// check if httpClient exists or not
+	if c.httpClient == nil {
 		return fmt.Errorf(c.ErrorMsg)
 	}
 
 	req = req.WithContext(ctx)
```
Comment (heiko-koehler) [inline internal/client.go:405]: That's interesting. You use request at the time API is invoked not when request is created. Should you pass context at all to functions which create request?

File: internal/client.go [hunk 6]
```diff
@@ -16,152 +16,207 @@ import (
...
-// RequestCompletionCallback defines the type of the request callback function
-type RequestCompletionCallback func(*http.Request, *http.Response, interface{})
+type ClientOption func(*Client) error
 
-// NewClient returns a wrapper around http/https (as per isHTTP flag) client with additions of proxy & session_auth if given
-func NewClient(credentials *prismgoclient.Credentials, userAgent string, absolutePath string, isHTTP bool) (*Client, error) {
-	if userAgent == "" {
-		return nil, fmt.Errorf("userAgent argument must be passed")
+// WithLogger sets the logger for the httpClient
+func WithLogger(logger *logr.Logger) ClientOption {
```
Comment (heiko-koehler) [inline internal/client.go:71]: Does it make sense to associate logger with all its variables to client as a whole? go-logr can attach Logger to context.Context. See https://pkg.go.dev/github.com/go-logr/logr#NewContext

Individual function extract logger from context then. See https://pkg.go.dev/github.com/go-logr/logr#FromContext

Logger is purely optional in context. A null logger is returned by FromContextOrDiscard() if not set.
Attaching logger to context becomes more necessary as call hierarchy gets deeper and each level attaches additional values to logger (in context).

File: internal/client.go (hunk 3 above)
Comment (thunderboltsid) [reply internal/client.go:188]: Tbh, I don't like the sessionauth logic being present in the client constructor. I'd rather split it out into its own method. We inherited this client from the terraform repo and this PR is one of the many refactorings that we'll have to do to whip it into shape.

File: internal/client.go (hunk 2 above)
Comment (thunderboltsid) [reply internal/client.go:211]: This is just request creation, there's no need for passing context in the parameter; will remove it altogether.

File: internal/client.go (hunk 5 above)
Comment (thunderboltsid) [reply internal/client.go:405]: No, removed them alotgether.

File: internal/client.go (hunk 6 above)
Comment (heiko-koehler) [reply internal/client.go:71]: We can defer this until need arises.

**Review by heiko-koehler — DISMISSED** [review]

File: fc/fc_service_test.go [hunk 7]
```diff
@@ -5,30 +5,34 @@ import (
...
 
-func setup() (*http.ServeMux, *internal.Client, *httptest.Server) {
+func setup(t *testing.T) (*http.ServeMux, *internal.Client, *httptest.Server) {
 	mux := http.NewServeMux()
 	server := httptest.NewServer(mux)
-	c, _ := internal.NewClient(&prismgoclient.Credentials{
+	creds := &prismgoclient.Credentials{
 		URL:      "https://10.2.242.13:9440",
 		Username: "admin",
 		Password: "Nutanix.123",
```
Comment (tuxtof) [inline fc/fc_service_test.go:25]: there is clear password here

Comment (thunderboltsid) [reply fc/fc_service_test.go]: fixed

**Review by tuxtof — APPROVED** [review]

**Review by yannickstruyf3 — APPROVED** [review]

**Pull request merged**